test/integration/fake_upstream.cc

```diff
@@ -73,22 +73,26 @@ void FakeStream::decodeMetadata(Http::MetadataMapPtr&& metadata_map_ptr) {
 }
 
 void FakeStream::encode100ContinueHeaders(const Http::HeaderMapImpl& headers) {
-  std::shared_ptr<Http::HeaderMapImpl> headers_copy(
+  std::unique_ptr<Http::HeaderMapImpl> headers_copy(
```

Comment:
Can you add a comment here and below on why we are doing this dance? It seems suboptimal and I'm unsure why the old code didn't work with shared_ptr correctly handling deletion at the right time?

Reply:
What comment would you like to see, exactly?

Reply:
Why it is that shared_ptr does not work. It's not clear to me.

Reply:
It's not thread-safe, but I don't know why either (...or whether it's expected to be, to be honest).

@lizan any ideas?

Reply:
shard_ptr is partially thread-safe, it does guard on the refcount part but not the holding object. But TBH I couldn't explain why exactly this case flags TSAN with libc++. I think we're in one of the not-well supported case listed [here](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual#faq), statically linking lib(std)c++, and not using instrumented libc++.

An alternative way to do this is use the copyable `TestHeaderMapImpl` and forget about smart pointers for the sake of simplicity. This path is not performance critical anyway so I guess that's fine too.

Reply:
I don't think it actually fix anything other than enforce the header free at the posted thread.
If this does solve the race, I would suggest restoring the shared_ptr above and write as below to transfer the ownership to posted thread.
```
parent_.connection().dispatcher().post([this, headers = std::move(headers_copy)]() -> void {
    encoder_.encode100ContinueHeaders(*headers);
  });
```

Reply:
@silentdai that code doesn't fix the issue.

@mattklein123 how do you want to proceed?

Reply:
@silentdai I tried it before commenting #7927, as I said it is likely a false positive of TSAN with non instrumented stdlib so we're not fixing anything really

Reply:
@PiotrSikora this "fix" doesn't make any sense to me, so it must be a bug in the sanitizer. I'm fine doing this, especially in test code, but can you put a comment that links back to this issue and perhaps says that it is probably a bug in the fuzzer, etc.?

Reply:
The thing is, there is enough articles about and hacks for thread-safety of `std::shared_ptr` passed to lambda that I'm not convinced that the old code is supposed to work.

I've added neutral comment about the issue, let me know if it works for you.

Reply:
As long as the shared_ptr is being copied, and not captured by reference, this should be safe, and must be a bug in either the compiler, std library, or the fuzzer. I will take a look at the comment.